app/controllers/auth/actions/AuthAction.scala

```diff
@@ -0,0 +1,209 @@
+            requests.AuthenticatedRequest(
+              userId = externalId,
+              agentRef = agentRef,
+              saUtr = saUtr.map(SaUtr),
+              nino = nino.map(Nino(_)),
+              isAgentActive = isAgentActive,
+              confidenceLevel = confidenceLevel,
+              credentials = credentials,
+              request = request
+            )
+          (agentRef.isDefined, isAgentActive) match {
+            case (true, false) =>
+              Future.successful(Left(Redirect(controllers.routes.ErrorController.notAuthorised)))
+            case (true, true)  =>
+              agentTokenCheck(request, newRequest)
+            case _             =>
+              pertaxAuthService.authorise[A, Request[A]](request).map {
+                case Some(r) => Left(r)
+                case _       => Right(newRequest)
+              }
+          }
+        case _                                                                                              => throw new RuntimeException("Can't find credentials for user")
+      } recover { case _: NoActiveSession =>
+      lazy val ggSignIn    = appConfig.loginUrl
+      lazy val callbackUrl = appConfig.loginCallback
+      Left(
+        Redirect(
+          ggSignIn,
+          Map(
+            "continue_url" -> Seq(callbackUrl),
+            "origin"       -> Seq(appConfig.appName)
+          )
+        )
+      )
+    }
+
+  private def agentInfo(enrolments: Set[Enrolment]): (Option[Uar], Boolean) =
+    enrolments
+      .find(_.key == "IR-SA-AGENT")
+      .map { enrolment =>
+        val d = enrolment.identifiers
+          .find(id => id.key == "IRAgentReference")
+          .map(key => Uar(key.value))
+        Tuple2(d, enrolment.isActivated)
+      }
+      .getOrElse(Tuple2(None, false))
+
+  private def citizenDetailsCheck[A](request: AuthenticatedRequest[A])(implicit
+    hc: HeaderCarrier
+  ): Future[AuthenticatedRequest[A]] =
+    (request.nino, request.saUtr, request.isAgent) match {
+      case (Some(nino), None, false) =>
+        getSAUTRFromCitizenDetails(nino).map {
+          case retrievedSAUtr @ Some(_) => request.copy(saUtr = retrievedSAUtr)
+          case None                     => request
+        }
+      case _                         => Future.successful(request)
+    }
+
+  private def getSAUTRFromCitizenDetails(nino: Nino)(implicit hc: HeaderCarrier): Future[Option[SaUtr]] =
+    citizenDetailsService.getMatchingDetails(nino.nino).map {
+      case SucccessMatchingDetailsResponse(matchingDetails) =>
+        matchingDetails.saUtr match {
+          case Some(_) => matchingDetails.saUtr
+          case _       => None
+        }
+      case FailedMatchingDetailsResponse                    => None
+    }
+
+  private def notAuthorisedPage: Result      = Redirect(controllers.routes.ErrorController.notAuthorised)
+  private def serviceUnavailablePage: Result = Redirect(controllers.routes.ErrorController.serviceUnavailable)
```

Comment:
Don't redirect for server error. It masks the source of the issue.

Reply:
Not sure what you mean: the service unavailable page is being displayed because the token is not present for an agent. So it's not a server error - it's a user error presumably

Reply:
As per discussion I've reverted to notAuthorisedPage